pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -475,6 +487,28 @@ func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
 	return lastErr
 }
 
+// Provides a simple SRV lookup to join a memberlist cluster w/o knowning members' addresses upfront.
+// Note: Overwrites flag `memberlist.join` if provided in addition to `memberlist.join-service`.
+func (m *KV) lookupMembers(srv string) error {
```

Comment:
I'm wondering if we could reuse https://cortexmetrics.io/docs/configuration/arguments/#dns-service-discovery instead of having another way to configure service discovery. The DNS service discovery we're currently using is based on Thanos one: you can see an example of its usage at `pkg/querier/blocks_store_balanced_set.go`.

Would it make sense to use it for gossip too?

Reply:
@pracucci & @pstibrany Giving this a thought for the last couple of days, I believe it is a good idea to expand memberlist joins to use the same DNS service discovery in addition to our `joinMembers` flag. Although a simple `net.lookupSRV` is enough the use cases provided by Thanos' DNS provider make sense also for memberlists:
- Joining a cluster means resolving the addresses of all members upfront. However, leaving the resolution mechanism to a third-party dependency is good from a separation of concerns perspective and adds testability. OTOH my simple `net.lookupSRV` would need a function pointer handle to make it testable, that's not a big deal though.
- Re-Joining the cluster - e.g. periodically - should work the same as joining anew with the only difference of re-establishing connections.

AFAICS, the DNS service discovery mechanism is free of any caching of members and in addition to that the ultimate list of which members to connect to is maintained on memberlist side. In addition, reading through the hashicorp lib I don't see a blocker regarding how to resolve member addresses. WDYT?

@pracucci FWIW, using the DNS service disovery is not bound to DNS caching timeouts or so like that would demand to rejoin the cluster based on these timeouts right?

Reply:
> AFAICS, the DNS service discovery mechanism is free of any caching of members and in addition to that the ultimate list of which members to connect to is maintained on memberlist side. In addition, reading through the hashicorp lib I don't see a blocker regarding how to resolve member addresses. WDYT?

The way the Thanos DNS provider works is that `Addresses()` returns the result of the last `Resolve()`. For this specific use case you could just call `Resolve()` and then `Address()` in sequence.

> FWIW, using the DNS service disovery is not bound to DNS caching timeouts or so like that would demand to rejoin the cluster based on these timeouts right?

I'm not sure about the question. The DNS service discovery just uses the Go's DNS Lookup under the hood, so there's real difference compared to your current implementation, except that it would be more generic and we would reuse the same mechanism we have.

WDYT?

Reply:
My concern question is rather a concern of the sort: Using the thanos DNS provider do I need to implement re-joins based on DNS TTLs?

Reply:
I wouldn't bring TTLs into the question here.